### PR TITLE
GET-141 scrape salary range

### PR DESCRIPTION
**Pull request by smoothcontract**
### Context
Salary range was previously rendered on the fly using Nokogiri to parse the job profile body content. This was too slow when viewing a large number of job profiles on the search results page.

### Changes proposed in this pull request
Added new attributes to job profile for salary min and max values. These are populated when scraping the job profile, and then used within the decorator. Also cleaned up a little related/dead code.

As the salary values are held as integers, they need to be reformatted back to currency values. I've changed the default locale to `en-GB` to accomodate this, otherwise we would need to specify currency symbol every time. Open to other approaches though.

Search results page (for all 857 jobs) is now nice and quick:

```
  Rendering explore_occupations/results.html.erb within layouts/application
  Rendered shared/search/_results_form.html.erb (0.8ms)
  Rendered shared/_contact_us.html.erb (0.5ms)
  Rendered explore_occupations/results.html.erb within layouts/application (408.6ms)
  Rendered shared/navigation/_main_header.html.erb (0.7ms)
Completed 200 OK in 755ms (Views: 429.6ms | ActiveRecord: 124.8ms)
```

### Guidance to review
Job profiles need to be re-scraped by running `bundle exec rails data_import:scrape_job_profiles` - otherwise all job profiles will display `Varied` for salary range. It's not necessary to run the other scraper steps just for this.

### CODE REVIEW

File: app/models/job_profile.rb [hunk 1]
```diff
@@ -18,37 +18,8 @@ def scrape
     scraper = JobProfileScraper.new
     scraped = scraper.scrape(source_url)
 
-    update(name: scraped['title'], description: scraped['description'], content: scraped['body'])
-    self.skills = Skill.import(scraped['skills'])
-  end
-
-  def salary
```
Comment (smoothcontract) [inline app/models/job_profile.rb]: This is all dead code - and breaks because the constants are not defined here. Looks like it was a remnant from introducing the decorator on a previous PR (where these methods can be found).

File: app/models/job_profile.rb [hunk 2]
```diff
@@ -18,37 +18,8 @@ def scrape
...
-             .capitalize
-  end
-
-  private
-
-  def html_body
-    @html_body ||= Nokogiri::HTML(content)
+    self.name = scraped.delete('title')
+    self.skills = Skill.import scraped.delete('skills')
+    update!(scraped)
```
Comment (smoothcontract) [inline app/models/job_profile.rb]: It was getting a little tedious to specify each attribute here, so other than `title` (which can't be named `name` as it breaks Wombat), and `skills` which need to be handled differently this now just does a mass assignment of all scraped attributes.

File: app/decorators/job_profile_decorator.rb [hunk 3]
```diff
@@ -76,7 +71,7 @@ def section(xpath: nil)
   private
 
   def html_body
-    @html_body ||= Nokogiri::HTML(__getobj__ .content)
+    @html_body ||= Nokogiri::HTML(content)
```
Comment (smoothcontract) [inline app/decorators/job_profile_decorator.rb]: Works just as well without the `__getobj__` stuff and cleanly delegates `content` to the underlying model.

File: app/scrapers/job_profile_scraper.rb [hunk 4]
```diff
@@ -3,6 +3,15 @@ class JobProfileScraper
...
   title css: 'h1.heading-xlarge'
   description css: '.column-desktop-two-thirds p'
-  body 'css=body', :html
+
+  salary_min css: '#Salary p.dfc-code-jpsstarter/text()' do |min|
+    min.gsub(/\D/, '').to_i if min =~ /\d/
+  end
+
+  salary_max css: '#Salary p.dfc-code-jpsexperienced/text()' do |max|
+    max.gsub(/\D/, '').to_i if max =~ /\d/
```
Comment (smoothcontract) [inline app/scrapers/job_profile_scraper.rb]: Not sure if there's a nicer way to implement this. I want a `nil` value returned if the salary is not present with some digits, or just the numeric value if it is.

File: config/locales/en.yml [hunk 5]
```diff
@@ -1,4 +1,4 @@
-en:
+en-GB:
```
Comment (smoothcontract) [inline config/locales/en.yml]: 🤔

File: app/scrapers/job_profile_scraper.rb [hunk 6]
```diff
@@ -3,6 +3,15 @@ class JobProfileScraper
...
+  end
+
+  content 'css=body', :html
```
Comment (smoothcontract) [inline app/scrapers/job_profile_scraper.rb]: Renamed from `body` to match job profile attribute name

File: app/decorators/job_profile_decorator.rb [hunk 7]
```diff
@@ -20,13 +19,9 @@ class JobProfileDecorator < SimpleDelegator
   OTHER_ROUTES_SECTION_XPATH = "//section[contains(@class, 'job-profile-subsection') and contains (@id, 'otherroutes')]".freeze
 
   def salary_range
-    min_salary = html_body.xpath(SALARY_MIN_XPATH).children[0]
+    return 'Variable' unless salary_min && salary_max
```
Comment (tavy88) [inline app/decorators/job_profile_decorator.rb]: No big deal but I think testing if only min exists should be enough? Probably won't make much performance difference.

File: app/decorators/job_profile_decorator.rb [hunk 8]
```diff
@@ -1,8 +1,7 @@
 class JobProfileDecorator < SimpleDelegator
   include ActionView::Helpers::TagHelper
+  include ActionView::Helpers::NumberHelper
 
-  SALARY_MIN_XPATH = "//div[@id='Salary']//p[@class='dfc-code-jpsstarter']".freeze
-  SALARY_MAX_XPATH = "//div[@id='Salary']//p[@class='dfc-code-jpsexperienced']".freeze
   WORKING_HOURS_XPATH = "//div[@id='WorkingHours']//p[@class='dfc-code-jphours']".freeze
   WORKING_HOURS_PATTERNS_XPATH = "//div[@id='WorkingHoursPatterns']//p[@class='dfc-code-jpwpattern']".freeze
```
Comment (tavy88) [inline app/decorators/job_profile_decorator.rb]: I think you can remove these 2 as well now?

File: app/models/job_profile.rb [hunk 9]
```diff
@@ -18,37 +18,8 @@ def scrape
...
-             .strip
-             .capitalize
-  end
-
-  private
-
-  def html_body
-    @html_body ||= Nokogiri::HTML(content)
+    self.name = scraped.delete('title')
+    self.skills = Skill.import scraped.delete('skills')
```
Comment (tavy88) [inline app/models/job_profile.rb]: Could we do `Skill.import scraped.delete('skills')` on a separate method? This line seems to be doing 2 things.

File: app/models/job_profile.rb [hunk 10]
```diff
@@ -18,37 +18,8 @@ def scrape
...
-  end
-
-  def working_hours_patterns
-    html_body.xpath(WORKING_HOURS_PATTERNS_XPATH)
-             .children[0]
-             .text
-             .strip
-             .capitalize
-  end
-
```
Comment (tavy88) [inline app/models/job_profile.rb]: How are specs still passing if you are removing these methods? A bit confused... can't see them as attributes on the job_profile?

Comment (smoothcontract) [reply app/models/job_profile.rb]: This code is not called from anywhere else. I suspect you decided to move it to the decorator at some point and forgot to remove from the model.

File: app/decorators/job_profile_decorator.rb (hunk 7 above)
Comment (smoothcontract) [reply app/decorators/job_profile_decorator.rb]: If max is nil then it still makes more sense to show Various I think. We’re scraping an external site so best not to bake in too many assumptions about when these values get populated.

File: app/decorators/job_profile_decorator.rb (hunk 8 above)
Comment (smoothcontract) [reply app/decorators/job_profile_decorator.rb]: No, they’re still used in the decorator. I removed the working hours methods and other dead code from the model. The xpath for salary isn’t required now as that’s implemented in the job profile scraper.

File: app/models/job_profile.rb (hunk 10 above)
Comment (tavy88) [reply app/models/job_profile.rb]: Ah sorry, you're right. I thought I was looking at the decorator class.

File: app/decorators/job_profile_decorator.rb (hunk 8 above)
Comment (tavy88) [reply app/decorators/job_profile_decorator.rb]: Yep sorry I was misled by the removals in the job profile class.

File: app/decorators/job_profile_decorator.rb (hunk 7 above)
Comment (tavy88) [reply app/decorators/job_profile_decorator.rb]: Yeah but I don't think you'll have min and ni max or vice versa. If you have variable the min will likely miss. And there's no point checking both upper and lower limits. But again it's just an optional thing.

File: config/application.rb [hunk 11]
```diff
@@ -28,5 +28,7 @@ class Application < Rails::Application
     # the framework and any gems in your application.
 
     config.exceptions_app = routes
+
+    config.i18n.default_locale = :'en-GB'
```
Comment (cwrw) [inline config/application.rb]: I'm surprised we need the gem for this, I thought it would come with Rails: https://guides.rubyonrails.org/i18n.html#setup-the-rails-application-for-internationalization
unless they don't have GB?

File: app/decorators/job_profile_decorator.rb [hunk 12]
```diff
@@ -3,10 +3,9 @@
 # which should remove the need to disable rubocop rules here.
 class JobProfileDecorator < SimpleDelegator
   include ActionView::Helpers::TagHelper
+  include ActionView::Helpers::NumberHelper
```
Comment (cwrw) [inline app/decorators/job_profile_decorator.rb]: is this needed for precision?

File: app/scrapers/job_profile_scraper.rb (hunk 4 above)
Comment (cwrw) [reply app/scrapers/job_profile_scraper.rb]: Because we need `nil` I think this is ok

File: spec/scrapers/job_profile_scraper_spec.rb [hunk 13]
```diff
@@ -21,8 +21,16 @@
       expect(scraped['description']).to match 'organising meetings'
     end
 
+    it 'parses job profile minimum salary' do
```
Comment (cwrw) [inline spec/scrapers/job_profile_scraper_spec.rb]: can we add more tests to check for nil values, values with and without other characters?

File: config/locales/en.yml (hunk 5 above)
Comment (cwrw) [reply config/locales/en.yml]: seems ok according to https://guides.rubyonrails.org/i18n.html#configure-the-i18n-module, what are your current thoughts?

File: app/decorators/job_profile_decorator.rb (hunk 12 above)
Comment (smoothcontract) [reply app/decorators/job_profile_decorator.rb]: It's for `number_to_currency` helper

File: config/application.rb (hunk 11 above)
Comment (smoothcontract) [reply config/application.rb]: Annoyingly GB isn't included by default with Rails and is only included in this gem. `UNACCEPTABLE!`.

File: config/locales/en.yml (hunk 5 above)
Comment (smoothcontract) [reply config/locales/en.yml]: It works fine, and we're targettng GB only so I'm happy with it.

File: app/decorators/job_profile_decorator.rb (hunk 7 above)
Comment (smoothcontract) [reply app/decorators/job_profile_decorator.rb]: Proposing to leave this as is

File: app/models/job_profile.rb (hunk 9 above)
Comment (smoothcontract) [reply app/models/job_profile.rb]: I don't think it would be any clearer TBH. Proposing to leave this as is.

File: spec/scrapers/job_profile_scraper_spec.rb (hunk 13 above)
Comment (smoothcontract) [reply spec/scrapers/job_profile_scraper_spec.rb]: I've added additional test cases for minimum salary; maximum salary is pretty much identical so just left the happy path test for that one.

File: spec/scrapers/job_profile_scraper_spec.rb [hunk 14]
```diff
@@ -1,32 +1,87 @@
...
+  describe 'salary_max' do
+    it 'parses job profile maximum salary' do
+      expect(scraped['salary_max']).to eq 30_000
+    end
+  end
+
+  describe 'salary_min' do
+    let(:fake_page) { Mechanize::Page.new nil, nil, body, 200, scraper.mechanize }
+
+    around do |example|
```
Comment (cwrw) [inline spec/scrapers/job_profile_scraper_spec.rb]: Can you explain what those steps are doing as Im not sure I follow?

Comment (smoothcontract) [reply spec/scrapers/job_profile_scraper_spec.rb]: Wombat is a bit odd in parts. Here I’m building a mechanize page with defined body content rather than allowing it to be fetched from vcr cassette. 

The ‘metadata.page’ part is actually an assignment - method name is slightly weird as it forms part of the DSL.

So this swaps out the page for some known content, runs the spec and then resets to nil. The last step is crucial as page is lazily fetched otherwise and can pollute subsequent tests.
